File: src/main/java/main/commands/slashcommands/commands/management/LanguageCommand.java
package main.commands.slashcommands.commands.management;

import main.constants.Permission;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.component.interactions.selectionmenu.StringSelectMenuOption;
import main.utils.component.interactions.selectionmenu.StringSelectionMenuBuilder;
import main.utils.json.requestchannel.RequestChannelConfig;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocale;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.jetbrains.annotations.NotNull;

public class LanguageCommand extends AbstractSlashCommand {
    @Override
    protected void buildCommand() {
        setCommand(
                getBuilder()
                        .setName("language")
                        .setDescription("Change the bot's language!")
                        .addOptions(CommandOption.of(
                                OptionType.STRING,
                                "lang",
                                "The specific language to switch to",
                                false,
                                RobertifyLocale.getAvailableLanguages().stream()
                                        .map(RobertifyLocale::name)
                                        .toList()
                        ))
                        .setAdminOnly()
                        .build()
        );
    }

    @Override
    public String getHelp() {
        return null;
    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        if (!checks(event)) return;

        final var guild = event.getGuild();
        final var localeManager = LocaleManager.getLocaleManager(guild);
        final var lang = event.getOption("lang");
        if (lang == null) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LanguageCommandMessages.LANGUAGE_EMBED_DESC, Pair.of("{language}", localeManager.getLocale().getLocalName() + " " + localeManager.getLocale().getFlag())).build())
                    .addActionRow(StringSelectionMenuBuilder.of(
                            "languagemenu",
                            localeManager.getMessage(RobertifyLocaleMessage.LanguageCommandMessages.LANGUAGE_SELECT_MENU_PLACE_HOLDER),
                            Pair.of(1,1),
                            RobertifyLocale.getAvailableLanguages().stream()
                                    .map(locale -> StringSelectMenuOption.of(
                                            locale.getName(),
                                            "languagemenu:" + locale.name().toLowerCase(),
                                            Emoji.fromUnicode(locale.getFlag())
                                    ))
                                    .toList()
                    ).build())
                    .setEphemeral(true)
                    .queue();
        } else {
            final var newLocale = RobertifyLocale.parse(lang.getAsString());
            localeManager.setLocale(newLocale);
            new RequestChannelConfig(guild).updateAll();
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LanguageCommandMessages.LANGUAGE_CHANGED, Pair.of("{language}", newLocale.getLocalName() + " " + newLocale.getFlag()))
                    .build())
                    .setEphemeral(true)
                    .queue();
        }
    }
    
    
    @Override
    public void onStringSelectInteraction(@NotNull StringSelectInteractionEvent event) {
        final var selectionMenu = event.getSelectMenu();
        if (!selectionMenu.getId().startsWith("languagemenu")) return;

        final var newLocale = RobertifyLocale.parse(event.getSelectedOptions().get(0).getValue().split(":")[1]);
        final var guild = event.getGuild();
        final var member = event.getMember();

        if (!GeneralUtils.hasPerms(guild, member, Permission.ROBERTIFY_ADMIN)) {
            event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(
                    guild,
                    RobertifyLocaleMessage.GeneralMessages.NO_MENU_PERMS
                            ).build()
                    )
                    .setEphemeral(true)
                    .queue();
            return;
        }

        event.deferReply().queue();

        LocaleManager.getLocaleManager(guild).setLocale(newLocale);
        new RequestChannelConfig(guild).updateAll();
        event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LanguageCommandMessages.LANGUAGE_CHANGED, Pair.of("{language}", newLocale.getLocalName() + " " + newLocale.getFlag()))
                .build())
                .setEphemeral(true)
                .queue();
    }
}

File: src/main/kotlin/main/commands/slashcommands/management/LanguageCommand.kt
package main.commands.slashcommands.management

import main.main.Robertify
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.selectionmenu.StringSelectMenuOption
import main.utils.component.interactions.selectionmenu.StringSelectionMenuBuilder
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.json.requestchannel.RequestChannelConfig
import main.utils.locale.LocaleManager
import main.utils.locale.RobertifyLocale
import main.utils.locale.messages.LanguageCommandMessages
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.emoji.Emoji
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent
import net.dv8tion.jda.api.sharding.ShardManager

class LanguageCommand : AbstractSlashCommand(
    SlashCommand(
        name = "language",
        description = "Change the bot's language.",
        options = listOf(
            CommandOption(
                name = "language",
                description = "The specific language to switch to.",
                choices = RobertifyLocale.availableLanguages.map { it.name },
                required = false
            )
        ),
        adminOnly = true
    )
) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val guild = event.guild!!
        val localeManager = LocaleManager[guild]
        val language = event.getOption("language")?.asString

        if (language == null) {
            event.replyEmbed {
                embed(
                    LanguageCommandMessages.LANGUAGE_EMBED_DESC,
                    Pair("{language}", "${localeManager.locale.name} ${localeManager.locale.flag}")
                )
            }.setActionRow(StringSelectionMenuBuilder(
                _name = "languagemenu",
                placeholder = localeManager.getMessage(LanguageCommandMessages.LANGUAGE_SELECT_MENU_PLACE_HOLDER),
                range = Pair(1, 1),
                _options = RobertifyLocale.availableLanguages.map { locale ->
                    StringSelectMenuOption(
                        label = locale.localName,
                        value = "languagemenu:${locale.name.lowercase()}",
                        emoji = Emoji.fromUnicode(locale.flag)
                    )
                }
            ).build())
                .setEphemeral(true)
                .queue()
        } else {
            val newLocale = RobertifyLocale.parse(language)
            localeManager.locale = newLocale
            RequestChannelConfig(guild).updateAll()
            event.replyEmbed {
                embed(
                    LanguageCommandMessages.LANGUAGE_CHANGED,
                    Pair("{language}", newLocale.localName)
                )
            }.setEphemeral(true)
                .queue()
        }
    }

    override suspend fun onStringSelectInteraction(event: StringSelectInteractionEvent) {
        val selectMenu = event.selectMenu
        if (selectMenu.id?.startsWith("languagemenu") != true)
            return

        val guild = event.guild!!

        event.deferReply(true).queue()

        val (_, locale) = event.selectedOptions.first().value.split(":")
        val newLocale = setLocale(guild, locale)

        event.hook.sendEmbed(guild) {
            embed(
                LanguageCommandMessages.LANGUAGE_CHANGED,
                Pair("{language}", newLocale.localName)
            )
        }.queue()
    }

    suspend fun setLocale(guild: Guild, locale: String, shardManager: ShardManager = Robertify.shardManager): RobertifyLocale {
        val newLocale = RobertifyLocale.parse(locale)
        LocaleManager[guild].locale = newLocale
        RequestChannelConfig(guild, shardManager).updateAll()
        return newLocale
    }
}